src/codegen/aggregation.cpp

```diff
@@ -39,15 +43,90 @@ void Aggregation::Setup(
         const type::Type count_type{type::BigInt::Instance()};
         uint32_t storage_pos = storage_.AddType(count_type);
 
+        // add hash table if distinct
+        if (agg_term.distinct) {
+          // Register the hash-table instance in the runtime state
+          auto hash_table_id = runtime_state.RegisterState(
+              "agg_hash" + std::to_string(source_idx),
+              OAHashTableProxy::GetType(codegen));
+
+          // Prepare the hash keys
+          // if not global, start with grouping keys, then add aggregation key
+          std::vector<type::Type> key_type;
+          if (!IsGlobal()) {
+            key_type = grouping_ai_types;
+          }
+          key_type.push_back(agg_term.expression->ResultType());
+
+          // Create the hash table
+          // we don't need to save any values, so value_size is zero
+          auto hash_table = OAHashTable{codegen, key_type, 0};
+
+          hash_table_infos_.emplace_back(
+              std::pair<OAHashTable, RuntimeState::StateID>(
+                  std::move(hash_table), hash_table_id));
```

Comment:
This logic is copied in different case branches, can you extract it somehow? Also the cases are getting quite complex, consider doing multiple passes over the aggregates during setup.